rocAL Tensor Node augmentations

fiona-gladwin opened the pull request:
- Add tensor support for all node augmentations
- Add support to invoke tensor related OpenVX augmentations.
- Add changes in rocal_api_augmentations.h and rocal_api_augmentations.cpp file

fiona-gladwin committed:
Modify unit test wrt tensor API

fiona-gladwin committed:
Add brightness fixed support

fiona-gladwin committed:
Grayscale inputs fix

fiona-gladwin committed:
Unittest fix

Fix for Golden outputs
Add tests for tensor API in testScript

fiona-gladwin committed:
Minor change

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Create vector of cv matrix for multiple outputs

fiona-gladwin committed:
Minor fix

fiona-gladwin committed:
Tensor support for rotate and warp affine

fiona-gladwin committed:
Fix RMN tensor support

fiona-gladwin committed:
Tensor support - Flip, Fisheye, LensCorrection

fiona-gladwin committed:
Minor fix

Commit:
Change name of vx augmentations API

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Merge branch 'lk/rocalTensor' of https://github.com/LakshmiKumar23/MIVisionX into PR_tensor_unittest1

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Minor change - Remove out buffer allocation

fiona-gladwin committed:
Modify augmentation nodes wrt tensor support

fiona-gladwin committed:
Add changes in node files

fiona-gladwin committed:
Fix build issues temporarily

fiona-gladwin committed:
Add changes to color augmentations

fiona-gladwin committed:
Added Support for effect augmentations

fiona-gladwin committed:
Add tensor support crop

fiona-gladwin committed:
Add API to create crop tensor

fiona-gladwin committed:
Tensor support  geometric nodes

Resize_crop_mirror
Random crop

fiona-gladwin committed:
minor code change

fiona-gladwin committed:
Minor code clean up include files

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Remove additional copy

Fix warnings

fiona-gladwin committed:
Minor change - Variable name

fiona-gladwin committed:
Add changes in ssd random crop and crop resize

fiona-gladwin committed:
Code clean up and minor changes

geometric augmentations

fiona-gladwin committed:
Minor changes - Variable name

fiona-gladwin committed:
Add changes in param for mean, std_dev and mirror

fiona-gladwin committed:
Revert multiple cv matrix creation in unit test

fiona-gladwin committed:
Add number of outputs condition check

fiona-gladwin committed:
Change name of contrast params

fiona-gladwin committed:
Merge branch 'lk/rocalTensor' of https://github.com/LakshmiKumar23/MIVisionX into PR_tensor_node_augmentations

fiona-gladwin committed:
Revert formatting changes - node include files

fiona-gladwin committed:
Minor formatting changes

fiona-gladwin committed:
Align arguments wrt function declaration

fiona-gladwin committed:
Formatting changes api_augmentations.cpp

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Fix rocalCropResize - remove unused args

Fix copy, nop and ssd random crop
Minor formatting changes

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Add lambda to modify width and height in dims

fiona-gladwin committed:
Minor build fix

fiona-gladwin committed:
Formatting changes augmentations

fiona-gladwin committed:
Minor formatting changes - Variable names

And build fix

fiona-gladwin committed:
Align params wrt coding guidelines

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Change API name from vxRpp to vxExtRpp

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Minor change

fiona-gladwin committed:
Remove additional rocalFlip call

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Change ROCAL_NHWC to ROCAL_NONE in default arguments

fiona-gladwin committed:
Change argument name

fiona-gladwin committed:
Add fix for grayscale outputs

LakshmiKumar23 commented on rocAL/rocAL/include/api/rocal_api_augmentation.h:
```diff
@@ -33,7 +33,7 @@ THE SOFTWARE.
 /// \param is_output
 /// \return
-extern "C"  RocalTensor  ROCAL_API_CALL rocalSequenceRearrange(RocalContext p_context, RocalTensor p_input,
+extern "C" RocalTensor  ROCAL_API_CALL rocalSequenceRearrange(RocalContext p_context, RocalTensor p_input,
```
@fiona-gladwin All changes to header files should have comments like the ones we have added to vx_rpp extension. Please follow the same format as rocAL repo. All the changes implemented there for the documentation needs to be merged to this branch

fiona-gladwin replied:
Done. Added Doxygen comments for all augmentations following the format in rocAL repo

LakshmiKumar23 replied:
@fiona-gladwin I do not think it follows the format. Please refer to this PR. The inputs need to be ```[in]``` and so on. Please also merge with the TOT on rocAL repo, since that has the code needed to generate the files. Once it's merged we have to run that one line command to generate the file to make sure it looks ok. Please let me know if you have any questions

fiona-gladwin replied:
@LakshmiKumar23 In rocAL repo this is the format that is added for the comments hence we followed the same - [link](https://github.com/ROCmSoftwarePlatform/rocAL/blob/516cdc2e4738635487037c4e88f4434fd84281b7/rocAL/include/api/rocal_api_augmentation.h#L173)

Please let me know if we need to have comments similar to vx_rpp or rocAL repo and we can make changes accordingly

LakshmiKumar23 replied:
@fiona-gladwin https://github.com/ROCmSoftwarePlatform/rocAL/commit/6ed79977410f3a6bb03fff9972f0de52c6c7e39b#diff-3ee61a464ddcf9a70dd656c57a8af4355a28eb8820c1acaea25b153542f99109
like in this PR. Inputs/outputs/bidirectional are defined as ```[in] [out] [inout]``` respectively. 
I have made those changes in vx_rpp extensions. same for all header files in rpp and rocAL going forward

fiona-gladwin replied:
@LakshmiKumar23 In the above link I could see in rocal_api_augmentations.h the [in] and [out] specifiers are not present.
Should we add them?

LakshmiKumar23 replied:
@fiona-gladwin Sorry, wrong link https://github.com/ROCmSoftwarePlatform/rocAL/commit/6ed79977410f3a6bb03fff9972f0de52c6c7e39b#diff-d1172c0cf6822f490cd3502421e798f0fd01dd040bb8b3285df37ea92fb836eb

fiona-gladwin committed:
Modify unit test

Add support to check all augmentations with file source reader
Enable remianing augmentations in unit test

fiona-gladwin committed:
Add parameters description

fiona-gladwin committed:
Changed params description  wrt Doxygen

LakshmiKumar23 commented on utilities/rocAL/rocAL_unittests/testAllScripts.sh:
```diff
@@ -85,78 +85,78 @@ do
-        ./rocAL_unittests 2 "$coco_detection_path" "${output_path}Vignette_${rgb_name[$rgb]}_${device_name}" $width $height 38 $device $rgb 0 $display
+        ./rocAL_unittests 0 "$image_path" "${output_path}Gamma_${rgb_name[$rgb]}_${device_name}" $width $height 33 $device $rgb 0 $display
+        ./rocAL_unittests 0 "$image_path" "${output_path}Contrast_${rgb_name[$rgb]}_${device_name}" $width $height 34 $device $rgb 0 $display
+        ./rocAL_unittests 0 "$image_path" "${output_path}Vignette_${rgb_name[$rgb]}_${device_name}" $width $height 38 $device $rgb 0 $display
```
@fiona-gladwin Why are we using same case 0 for all test cases? In that case the comments are all misleading

shobana-mcw replied:
We have not given support for all readers in this PR yet. Since we want to check the functionalities of all augmentations currently this unit test will run and compare against Golden outputs using File source reader. 

In the readers PR we will revert back to our TOT unit test which tests all readers + augmentations.

LakshmiKumar23 commented on rocAL/rocAL/include/augmentations/geometry_augmentations/node_flip.h:
```diff
@@ -25,23 +25,22 @@ THE SOFTWARE.
\ No newline at end of file
+    ParameterVX<int> _horizontal, _vertical;
+    constexpr static int HORIZONTAL_RANGE[2] = {0, 1};
+    constexpr static int VERTICAL_RANGE[2] = {0, 1};
```
@fiona-gladwin Why has this value changed?

shobana-mcw replied:
RPP has been changed to take 2 separate params for Vertical and Horizontal flip to match DALI. So we need to pass 2 params.

fiona-gladwin committed:
Merge branch 'lk/rocalTensor' of https://github.com/LakshmiKumar23/MIVisionX into PR_tensor_node_augmentations

fiona-gladwin committed:
Minor fix

fiona-gladwin committed:
Modify API description

Mark all params as input
Add description for each augmentation
Add return type in description

fiona-gladwin committed:
Minor changes

rrawther commented on rocAL/rocAL/include/api/rocal_api_augmentation.h:
```diff
@@ -24,538 +24,902 @@ THE SOFTWARE.
-/// \param is_output
-/// \return
-extern "C"  RocalTensor  ROCAL_API_CALL rocalSequenceRearrange(RocalContext p_context, RocalTensor p_input,
+/*!
```
make sure this can generate doxygen documentation

rrawther commented on rocAL/rocAL/include/api/rocal_api_types.h:
```diff
@@ -155,7 +155,8 @@ enum RocalResizeScalingMode {
     ROCAL_SCALING_MODE_NOT_SMALLER = 2, // scales wrt to aspect ratio, so that resize width/height is not lesser than the specified size
-    ROCAL_SCALING_MODE_NOT_LARGER = 3   // scales wrt to aspect ratio, so that resize width/height does not exceed specified size
+    ROCAL_SCALING_MODE_NOT_LARGER = 3,   // scales wrt to aspect ratio, so that resize width/height does not exceed specified size
+    ROCAL_SCALING_MODE_MIN_MAX = 4 // scales wrt to aspect ratio, so that resize width/height does not exceed specified min and max size
```
why is this new mode introduced?

fiona-gladwin replied:
This mode has been introduced for mask training. MaskRCNN requires a different resize scaling logic that's not available in the existing modes, hence in order to follow that logic for mask training we have introduced this mode so the user can enable this instead of the generic resize logic

rrawther commented on rocAL/rocAL/include/augmentations/effects_augmentations/node_jitter.h:
```diff
@@ -25,18 +25,17 @@ THE SOFTWARE.
     JitterNode() = delete;
-    void init(int kernel_size);
-    void init(IntParam *kernel_size);
+    void init(int kernel_size, int seed);
```
why is the seed only there for this node?

fiona-gladwin replied:
RPP Jitter and Noise augmentations accepts seed as an input
Hence we have added them in rocAL as well

fiona-gladwin committed:
Fixing context release issue

fiona-gladwin committed:
Merge branch 'PR_tensor_unittest1' into PR_tensor_node_augmentations

fiona-gladwin committed:
Merge branch 'lk/rocalTensor' of https://github.com/LakshmiKumar23/MIVisionX into PR_tensor_node_augmentations

rrawther commented on rocAL/rocAL/include/augmentations/geometry_augmentations/node_random_crop.h:
```diff
@@ -25,12 +25,11 @@ THE SOFTWARE.
     RandomCropNode(const std::vector<Tensor *> &inputs, const std::vector<Tensor *> &outputs);
     RandomCropNode() = delete;
-    void init(float area, float aspect_ratio, float x_drift, float y_drift);
+    ~RandomCropNode();
```
we don't need destructor here unless you want to do something special here. Otherwise the base class destructor will be called

fiona-gladwin replied:
The destructors are added in all the crop related nodes.
The crop coordinates needs to be passed as vx_tensor hence we need to create a vx_tensor in the node and release it.
Hence the destructor is introduced in all crop related nodes to release the vx_tensor for the crop

fiona-gladwin replied:
Added destructor only in CropNode will be used by other derived crop related augmentations like RandomCropNode, CropMirrorNormalizeNode etc

rrawther commented on rocAL/rocAL/include/augmentations/geometry_augmentations/node_resize_crop_mirror.h:
```diff
@@ -22,30 +22,34 @@ THE SOFTWARE.
     ResizeCropMirrorNode() = delete;
-    void init(unsigned int crop_h, unsigned int crop_w, IntParam *mirror);
-    void init( FloatParam *crop_h_factor, FloatParam *crop_w_factor, IntParam *mirror);
+    ~ResizeCropMirrorNode();
```
please reme destructor

fiona-gladwin replied:
Done

rrawther commented on rocAL/rocAL/include/augmentations/geometry_augmentations/node_rotate.h:
```diff
@@ -25,26 +25,24 @@ THE SOFTWARE.
     void update_node() override;
 private:
     ParameterVX<float> _angle;
-    vx_array _dst_roi_width,_dst_roi_height;
```
how is the dst_roi handled?

fiona-gladwin replied:
We have a _dst_tensor_roi and _src_tensor_roi in the node which will handle the src and dst coordinates

rrawther commented on rocAL/rocAL/include/augmentations/node_ssd_random_crop.h:
```diff
@@ -59,11 +59,11 @@ class SeededRNG {
 public:
     SSDRandomCropNode(const std::vector<Tensor *> &inputs, const std::vector<Tensor *> &outputs);
     SSDRandomCropNode() = delete;
+    ~SSDRandomCropNode();
```
remove

fiona-gladwin replied:
Done

rrawther commented on rocAL/rocAL/include/augmentations/node_ssd_random_crop.h:
```diff
@@ -79,10 +79,7 @@ class SSDRandomCropNode : public Node
-    std::vector<uint> _crop_width_val, _crop_height_val, _x1_val, _y1_val, _x2_val, _y2_val;
-    // unsigned int _dst_width, _dst_height;
-    std::vector<uint32_t> in_width, in_height;
+    std::vector<uint> _crop_width_val, _crop_height_val, _x1_val, _y1_val;
```
I would suggest x1, y1, width, height order instead

fiona-gladwin replied:
Done

rrawther commented on rocAL/rocAL/include/pipeline/node.h:
```diff
@@ -41,8 +41,9 @@ class Node
-    void set_meta_data(pMetaDataBatch meta_data_info){ _meta_data_info = meta_data_info;}
+    void set_meta_data(pMetaDataBatch meta_data_info) { _meta_data_info = meta_data_info; }
     bool _is_ssd = false;
+    void create_crop_tensor(vx_tensor &crop_tensor, void **crop_coordinates);
```
why are we creating a crop_tensor in node.h. This will be created for all the nodes which are derived

fiona-gladwin replied:
- The crop_tensor is not derived for all nodes. As they are created and maintained only by the derived crop nodes.
- The crop nodes handles the allocation and release.
- Only the function to create a crop tensor is present in node.h since we wanted to keep it common to all crop nodes we have introduced the API and it's definition here.
- This function accepts the vx_tensor and the ptr by reference which is created in the crop nodes which is derived from base node class and allocates the tensor passed by the derived nodes.

rrawther commented on rocAL/rocAL/include/pipeline/node.h:
```diff
@@ -54,5 +55,6 @@ class Node
     size_t _batch_size;
     pMetaDataBatch _meta_data_info;
-    vx_array _src_roi_width, _src_roi_height; // TODO - To be removed after tensor changes in augmentation nodes.
+    vx_scalar _input_layout, _output_layout, _roi_type;
```
node accepts a vector of inputs and outputs. Can't we get the layout from the Tensor object pointer? Please change this

fiona-gladwin replied:
All our nodes currently process 1 input and 1 output
All the OpenVX RPP nodes also accept only 3 params i.e input_layout, output_layout and roiType

Hence we have added this as part of the base node class so we need not replicate in every derived nodes.

rrawther commented on rocAL/rocAL/include/pipeline/node.h:
```diff
@@ -54,5 +55,6 @@ class Node
     pMetaDataBatch _meta_data_info;
-    vx_array _src_roi_width, _src_roi_height; // TODO - To be removed after tensor changes in augmentation nodes.
+    vx_scalar _input_layout, _output_layout, _roi_type;
+    vx_array _src_roi_width, _src_roi_height;  // TODO - To be removed after tensor changes in augmentation nodes.
```
this has to be part of input and output tensor

rrawther commented on rocAL/rocAL/include/pipeline/tensor.h:
```diff
@@ -138,7 +138,8 @@ class TensorInfo {
     void set_tensor_layout(RocalTensorlayout layout) {
-        if(_layout != layout && _layout != RocalTensorlayout::NONE) {
+        if (layout == RocalTensorlayout::NONE) return;
+        if (_layout != layout && _layout != RocalTensorlayout::NONE) {  // If layout input and current layout's are different modify dims accordingly
```
remove the second condition for if which is implied because of line #141

fiona-gladwin replied:
The first condition checks if the input layout is None
Whereas the second condition checks if the current layout of the tensor is None

LakshmiKumar23 commented:
@fiona-gladwin please resolve merge conflicts. I did for PR32, but there are still some unresolved comments there too

rrawther commented on rocAL/rocAL/source/api/rocal_api_augmentation.cpp:
```diff
@@ -20,15 +20,37 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 #include "rocal_api.h"
 #include "image_source_evaluator.h"
 
+auto modify_dims_width_and_height = [](RocalTensorlayout tensor_layout, std::vector<size_t> &dims, size_t width, size_t height) {
```
why are we having this function here

fiona-gladwin replied:
Removed

rrawther commented on rocAL/rocAL/source/api/rocal_api_augmentation.cpp:
```diff
@@ -61,425 +83,416 @@ rocalSequenceRearrange(RocalContext p_context,
+        
+        // For the rotate node, user can create a tensor with a different width and height
+        std::vector<size_t> out_dims = output_info.dims();
+        modify_dims_width_and_height(output_info.layout(), out_dims, dest_width, dest_height);
```
we don't want to have this function defined in this API class. Probably you need to add another fun called calculate_dims_xx() in the tensor class which claculates dimensions based on width, height, layout etc. API definitions whould only have calls to add augmentations. Here I see lot of calls to ceate tensor and add roi

fiona-gladwin replied:
Done

rrawther commented on rocAL/rocAL/source/api/rocal_api_augmentation.cpp:
```diff
@@ -61,425 +83,416 @@ rocalSequenceRearrange(RocalContext p_context,
-        output_info.height(dest_height);
-
-        output = context->master_graph->create_image(output_info, is_output);
+        RocalTensorlayout op_tensor_layout = static_cast<RocalTensorlayout>(output_layout);
```
let us discuss this over a call

rrawther commented on rocAL/rocAL/source/api/rocal_api_data_loaders.cpp:
```diff
@@ -150,17 +184,14 @@ rocalJpegFileSourceSingleShard(
         INFO("Internal buffer size width = "+ TOSTR(width)+ " height = "+ TOSTR(height) + " depth = "+ TOSTR(num_of_planes))
 
-        std::vector<size_t> dims = {context->user_batch_size(), height, width, num_of_planes};
+        auto [tensor_layout, dims] = convert_color_format_to_layout(rocal_color_format, context->user_batch_size(), height, width, num_of_planes);
```
please remove convert_color_format function and have only one function.

fiona-gladwin replied:
These changes were a part of the previous PR and these changes were visible here as it was not merged before.
This file is not a part of this PR. Will be changed in the upcoming PR for readers and meta nodes.

rrawther commented on rocAL/rocAL/source/api/rocal_api_data_loaders.cpp:
```diff
@@ -106,6 +106,40 @@ auto convert_decoder_mode= [](RocalDecodeDevice decode_mode)
+    switch(color_format) {
+        case ROCAL_COLOR_RGB24:
+        case ROCAL_COLOR_BGR24: {
+            std::vector<size_t> dims_vector = {n, f, h, w, c};
```
rename to dims (for below fun as well)

fiona-gladwin replied:
This file is not a part of this PR. Will be changed in the upcoming PR for readers and meta nodes.

rrawther commented on rocAL/rocAL/source/api/rocal_api_data_loaders.cpp:
```diff
@@ -150,17 +184,14 @@ rocalJpegFileSourceSingleShard(
 
-        std::vector<size_t> dims = {context->user_batch_size(), height, width, num_of_planes};
+        auto [tensor_layout, dims] = convert_color_format_to_layout(rocal_color_format, context->user_batch_size(), height, width, num_of_planes);
         auto info  = TensorInfo(std::move(dims),
```
can we define another constructor for TensorInfo() which takes all the parameters and set it in one call instead of calling multiple functions

fiona-gladwin replied:
This file is not a part of this PR. Will be changed in the upcoming PR for readers and meta nodes.

rrawther commented on rocAL/rocAL/source/augmentations/color_augmentations/node_brightness.cpp:
```diff
@@ -24,30 +24,22 @@ THE SOFTWARE.
-    _node = vxExtrppNode_Brightness(_graph->get(), _inputs[0]->handle(), _src_tensor_roi, _outputs[0]->handle(), _alpha.default_array(), _beta.default_array(), in_layout_vx, out_layout_vx, roi_type_vx);
+    _alpha.create_array(_graph, VX_TYPE_FLOAT32, _batch_size);
+    _beta.create_array(_graph, VX_TYPE_FLOAT32, _batch_size);
+    _node = vxExtRppBrightness(_graph->get(), _inputs[0]->handle(), _src_tensor_roi, _outputs[0]->handle(), _alpha.default_array(), _beta.default_array(), _input_layout, _output_layout, _roi_type);
```
please have the layout extracted from the tensor class instead of having another variables for it. Plus here the node can take a vector of tensors

fiona-gladwin replied:
Added changes to extract the input layout, output layout and roi type from the input and output tensor in every augmentations node class instead of having it common in Node class

rrawther commented on rocAL/rocAL/source/augmentations/color_augmentations/node_exposure.cpp:
```diff
@@ -26,35 +26,28 @@ THE SOFTWARE.
-    _shift.create_array(_graph , VX_TYPE_FLOAT32, _batch_size);
-    // _node = vxExtrppNode_ExposurebatchPD(_graph->get(), _inputs[0]->handle(), _src_roi_width, _src_roi_height, _outputs[0]->handle(), _shift.default_array(), _batch_size);
+    _exposure_factor.create_array(_graph, VX_TYPE_FLOAT32, _batch_size);
+    _node = vxExtRppExposure(_graph->get(), _inputs[0]->handle(), _src_tensor_roi, _outputs[0]->handle(), _exposure_factor.default_array(), _input_layout, _output_layout, _roi_type);
```
same comment as before for layout. Applies to all such

fiona-gladwin replied:
Done

rrawther commented on rocAL/rocAL/source/augmentations/effects_augmentations/node_rain.cpp:
```diff
@@ -31,47 +29,38 @@ RainNode::RainNode(const std::vector<Tensor *> &inputs, const std::vector<Tensor
-    // _node = vxExtrppNode_RainbatchPD(_graph->get(), _inputs[0]->handle(), _src_roi_width, _src_roi_height, _outputs[0]->handle(), _rain_value.default_array(), _rain_width.default_array(),
-                                                    // _rain_height.default_array(), _rain_transparency.default_array(), _batch_size);
-
+    _rain_value.create_array(_graph, VX_TYPE_FLOAT32, _batch_size);
```
it is better to combine all this parameters into a single tensor as discussed

fiona-gladwin committed:
Merge branch 'lk/rocalTensor' of https://github.com/LakshmiKumar23/MIVisionX into PR_tensor_node_augmentations

fiona-gladwin committed:
Minor change

fiona-gladwin committed:
Add single API to modify width and height based on dims

rrawther commented on rocAL/rocAL/source/augmentations/effects_augmentations/node_snp_noise.cpp:
```diff
@@ -21,42 +21,55 @@ THE SOFTWARE.
+    _salt_prob.create_array(_graph, VX_TYPE_FLOAT32, _batch_size);
+    _salt_value.create_array(_graph, VX_TYPE_FLOAT32, _batch_size);
+    _pepper_value.create_array(_graph, VX_TYPE_FLOAT32, _batch_size);
+    vx_scalar seed = vxCreateScalar(vxGetContext((vx_reference)_graph->get()), VX_TYPE_UINT32, &_seed);
```
why seed is only used for this node, but not other nodes

fiona-gladwin replied:
RPP requires seed to be passed specifically for Jitter and Noise hence we create a scalar for the seed and pass it to the VX_RPP node only for these 2 augmentations

rrawther commented on rocAL/rocAL/source/augmentations/geometry_augmentations/node_crop.cpp:
```diff
@@ -27,68 +27,79 @@ THE SOFTWARE.
-        THROW("Uninitialized destination dimension")
-
     _crop_param->create_array(_graph);
+    create_crop_tensor(_crop_tensor, &_crop_coordinates);
```
I am not sure why we need to create_crop_tensor here. I don't see it  is getting updated. Also we can't create crop tensor in node.h base class.

fiona-gladwin replied:
We are creating a crop tensor from handle(i.e _crop_coordinates) since VX_RPP accepts crop coordinates as vx_tensor and passes the handle to the RPP kernels
In the [update node](https://github.com/fiona-gladwin/MIVisionX/blob/ac0bed6de0df4bf2848ad5e3d333579c2fa00ffc/rocAL/rocAL/source/augmentations/geometry_augmentations/node_crop.cpp#L57) we are updating the values in the handle which will be reflected in the tensor as well.

The crop tensor is present only in the derived nodes only the API to create in present in the node.h as it can be used by other crop related nodes.

fiona-gladwin replied:
As discussed, Removed this function from node class.

- Introduced the create_crop_tensor only in CropNode
- For all crop related augmentations inheriting the class from CropNode
- Added the crop_tensor and crop_coordinates as protected variables so it can be accessed by all the derived class.
- The destructor is added only for crop nodes and removed from other crop related augmentations

rrawther commented on rocAL/rocAL/source/augmentations/geometry_augmentations/node_crop_mirror_normalize.cpp:
```diff
@@ -72,39 +75,15 @@ void CropMirrorNormalizeNode::create_node() {
-    if ((status = vxGetStatus((vx_reference)_crop_tensor)) != VX_SUCCESS)
-        THROW("Error: vxCreateTensorFromHandle(crop_tensor: failed " + TOSTR(status))
+        THROW(" vxAddArrayItems failed in the crop_mirror_normalize node (vxExtRppCropMirrorNormalize)  node: " + TOSTR(status) + "  " + TOSTR(status))   
+    create_crop_tensor(_crop_tensor, &_crop_coordinates);
```
same as previous comment

rrawther commented on rocAL/rocAL/source/pipeline/node.cpp:
```diff
@@ -23,11 +23,12 @@ THE SOFTWARE.
 {
-    if(!_node)
-        vxReleaseNode(&_node);
+    if (_src_tensor_roi) vxReleaseTensor(&_src_tensor_roi);
```
I would suggest roi_tensor as part of Tensor object since it can be tied to that.  For e.g, if a node has >1 inputs which one the _src_tensor_roi correspond to?

fiona-gladwin replied:
Done, Added the changes to have the vx_tensor for the ROI as part of tensor object
Removed the _src_tensor_roi and _dst_tensor_roi as discussed

rrawther commented on utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp:
```diff
@@ -383,7 +383,7 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
         scaling_node_name = get_scaling_mode(resize_scaling_mode, scale_mode);
         std::cerr<<" \n Interpolation_type_name " << interpolation_type_name;
-        std::cerr<<" \n Scaling_node_name " << scaling_node_name;
+        std::cerr<<" \n Scaling_node_name " << scaling_node_name << "\n ";
```
please use std::endl instead of "\n"

fiona-gladwin replied:
Done

rrawther commented on rocAL/rocAL/source/augmentations/geometry_augmentations/node_crop_resize.cpp:
```diff
@@ -38,51 +35,70 @@ void CropResizeNode::create_node()
 
-    // _node = vxExtrppNode_ResizeCropbatchPD(_graph->get(), _inputs[0]->handle(), _src_roi_width, _src_roi_height, _outputs[0]->handle(), _dst_roi_width,
-                                        //    _dst_roi_height, _crop_param->x1_arr, _crop_param->y1_arr, _crop_param->x2_arr, _crop_param->y2_arr, _batch_size);
+    create_crop_tensor(_crop_tensor, &_crop_coordinates);
```
see related comment for other nodes. Applies to all such

rrawther commented on rocAL/rocAL/source/augmentations/geometry_augmentations/node_resize_crop_mirror.cpp:
```diff
@@ -27,60 +27,84 @@ THE SOFTWARE.
 
+ResizeCropMirrorNode::~ResizeCropMirrorNode() {
+    if (_inputs[0]->info().mem_type() == RocalMemType::HIP) {
+#if ENABLE_HIP
```
this is not how we need to check if the pipeline is in cpu or device. For example ENABLE_HIP is default mode and user can create a pipeline on cpu or gpu. Need to use the device parameter of pipeline instead

fiona-gladwin replied:
We are making use of the mem_type() that is set to the tensor. If it is a HIP tensor then HIP mem is allocated and also removed for the ROI

rrawther commented on rocAL/rocAL/source/augmentations/node_ssd_random_crop.cpp:
```diff
@@ -182,28 +172,25 @@ void SSDRandomCropNode::update_node()
-
+        crop_dims[i].x1 = (crop_box.l) * input_roi[i].x2;
+        crop_dims[i].y1 = (crop_box.t) * input_roi[i].y2;
+        if(_inputs[0]->info().roi_type() == RocalROIType::XYWH) {
```
probably you need to define rocalROI as a union which can have x1,y1,x2, y2 or x,y,w,h so the code for setting and getting the values is more readable

fiona-gladwin replied:
Will issue a separate small PR for this change

rrawther reviewed (changes requested):
please address review comments

fiona-gladwin commented:
> @fiona-gladwin please resolve merge conflicts. I did for PR32, but there are still some unresolved comments there too

Done

fiona-gladwin committed:
Minor change

fiona-gladwin committed:
Add API in tensor info to modify strides

fiona-gladwin committed:
Merge branch 'lk/rocalTensor' of https://github.com/LakshmiKumar23/MIVisionX into PR_tensor_node_augmentations

fiona-gladwin committed:
Add ROI tensor changes

Remove tensor variable from node.h file
Add a vx_tensor for ROI in every tensor object
Add API to create the ROI tensor from handle

fiona-gladwin committed:
Add support to create vx_scalar for layout and ROI type in each node

fiona-gladwin committed:
Add changes to inherit crop augmentations from CropNode

Add API to create crop tensor in crop node
Remove destrutors in other crop nodes

fiona-gladwin committed:
Minor change

fiona-gladwin commented:
@rrawther and @LakshmiKumar23 Addressed all PR comments please check

LakshmiKumar23 commented:
@rrawther Good to merge?

fiona-gladwin committed:
Merge branch 'lk/rocalTensor' of https://github.com/LakshmiKumar23/MIVisionX into PR_tensor_node_augmentations

fiona-gladwin committed:
Fix include in video reader
